File: internal/modules/lbase64/base64_test.lua
local lust = require "lust"
local describe = lust.describe
local it = lust.it
local expect = lust.expect
local base64 = require("base64")

describe("base64 module", function()
    it("encode successfully", function ()
      local b = base64.encode("hello world")
      expect(b).to.be("aGVsbG8gd29ybGQ=")
    end)

    it("decode successfully", function ()
      local b, err  = base64.decode("aGVsbG8gd29ybGQ=")
      expect(err).to.be(nil)
      expect(b).to.be("hello world")
    end)

    it("encode_url successfully", function ()
      local b = base64.encode_url("https://www.example.com/bla?key1=val1&key2=val2#boo")
      expect(b).to.be("aHR0cHM6Ly93d3cuZXhhbXBsZS5jb20vYmxhP2tleTE9dmFsMSZrZXkyPXZhbDIjYm9v")
    end)

    it("decode_url successfully", function ()
      local b, err  = base64.decode_url("aHR0cHM6Ly93d3cuZXhhbXBsZS5jb20vYmxhP2tleTE9dmFsMSZrZXkyPXZhbDIjYm9v")
      expect(err).to.be(nil)
      expect(b).to.be("https://www.example.com/bla?key1=val1&key2=val2#boo")
    end)
end)

File: internal/modules/lbase64/base64_test.go
package lbase64

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

var stringTestData map[string]string = map[string]string{
	"hello world": "aGVsbG8gd29ybGQ=",
}

var urlTestData map[string]string = map[string]string{
	"https://www.example.com/bla?key1=val1&key2=val2#boo": "aHR0cHM6Ly93d3cuZXhhbXBsZS5jb20vYmxhP2tleTE9dmFsMSZrZXkyPXZhbDIjYm9v",
}

func TestEncodeString(t *testing.T) {
	for key, val := range stringTestData {
		assert.Equal(t, val, doEncodeString(key))
	}
}

func TestDecodeString(t *testing.T) {
	for key, val := range stringTestData {
		v, err := doDecodeString(val)
		assert.Nil(t, err)
		assert.Equal(t, key, string(v))
	}
}

func TestEncodeUrl(t *testing.T) {
	for key, val := range urlTestData {
		assert.Equal(t, val, doEncodeUrl(key))
	}
}

func TestDecodeUrl(t *testing.T) {
	for key, val := range urlTestData {
		v, err := doDecodeUrl(val)
		assert.Nil(t, err)
		assert.Equal(t, key, string(v))
	}
}
